### PR TITLE
[TIR] SplitHostDevice, handle subroutines

### PR DESCRIPTION
**Pull request by Lunderberg**
This PR refactors `SplitHostDevice` into three separate transformations.  Previously, `SplitHostDevice` would replace device regions with a `builtin::tvm_call_packed()` node to replace the extracted region.  After this PR, this process is performed in three separate steps.

1. `AnnotateDeviceRegion`: Annotate the regions that should be executed on another target.
2. `SplitHostDevice`: Extract the annotated region into an independent PrimFunc, with a `GlobalVar` to represent the call from into the new subroutine.
3. `LowerDeviceKernelLaunch`: For any subroutine call where the caller and callee are on different devices, replace with a device kernel launch.

### CODE REVIEW

**Comment by Lunderberg** [comment]
This PR is a subset of the functionality in https://github.com/apache/tvm/pull/14862.  As that PR has grown to a difficult-to-review size, I'm carving off independent chunks of it into separate PRs, after which #14862 can be rebased to a more manageable size.

In addition, this PR depends on https://github.com/apache/tvm/pull/14915 for the new unit tests, and should be rebased onto main after #14915 lands.

File: src/tir/transforms/lower_device_kernel_launch.cc [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+ * \brief Visitor class to collect device-side program information.
+ */
+class DeviceInfoCollector : public StmtVisitor {
+ public:
+  static KernelInfo Collect(const GlobalVar& gvar, const PrimFuncNode* func) {
+    DeviceInfoCollector collector;
+    collector.info_.target = [&]() -> Target {
+      auto target_attr = func->GetAttr<Target>(tvm::attr::kTarget).value();
+      bool is_host_func =
+          func->GetAttr<Bool>(tvm::tir::attr::kIsHostFunc).value_or(Bool(false))->value;
```
Comment (csullivan) [inline src/tir/transforms/lower_device_kernel_launch.cc:72]: Re. discussion offline, do we want to update to not rely on kIsHostFunc?

Comment (Lunderberg) [reply src/tir/transforms/lower_device_kernel_launch.cc]: Agreed, and updated.  The `SplitHostDevice` and `LowerDeviceKernelLaunch` no longer use the `kIsHostFunc` attribute at all.

**Comment by Lunderberg** [comment]
(Merged https://github.com/apache/tvm/pull/14950 into this PR to allow CI to run.  No conflicts expected after CI, so github's squash/merge will remove it from the final commit.)

**Review by csullivan — APPROVED** [review]
LGTM, thank you for the refactoring! Looking forward to the additional functionality having SplitHostDevice split out into separate steps will enable during lowering.

**Comment by tqchen** [comment]
there seems to be a regression caused by this PR, please look into it

**Comment by Lunderberg** [comment]
Found the issue.  It was a combination of this PR and #14945, which is why it wasn't caught by CI.  A full description of the issue, and a bugfix, are in PR #14982.

**Comment by tqchen** [comment]
A note that we should revisit some of the assumptions https://github.com/apache/tvm/issues/16237

**Pull request merged**